Use functions.json file for store

viveksyngh opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
Fixes: #618  #619 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tested on mac
```
./faas-cli-darwin store list --platform=arm64

FUNCTION  DESCRIPTION
NodeInfo  Get info about the machine that you'r...
sha512sum Generate shasums in 512 format

```
```
./faas-cli-darwin store list

FUNCTION                    DESCRIPTION
NodeInfo                    Get info about the machine that you'r...
Figlet                      Generate ASCII logos with the figlet CLI
SSL/TLS cert info           Returns SSL/TLS certificate informati...
Colorization                Turn black and white photos to color ...
Inception                   This is a forked version of the work ...
Have I Been Pwned           The Have I Been Pwned function lets y...
Face Detection with Pigo    Detect faces in images using the Pigo...
SentimentAnalysis           Python function provides a rating on ...
Business Strategy Generator Generates a Business Strategy (using ...
Tesseract OCR               This function brings OCR - Optical Ch...
Dockerhub Stats             Golang function gives the count of re...
QR Code Generator - Go      QR Code generator using Go
Nmap Security Scanner       Tool for network discovery and securi...
ASCII Cows                  Generate cartoons of ASCII cows
YouTube Video Downloader    Download YouTube videos as a function
OpenFaaS Text-to-Speech     Generate an MP3 of text using Google'...
nslookup                    Uses nslookup to return any IP addres...
Docker Image Manifest Query Query an image on the Docker Hub for ...
face-detect with OpenCV     Detect faces in images. Send a URL as...
Left-Pad                    left-pad on OpenFaaS
mememachine                 Turn any image into a meme.
```

```
./faas-cli-darwin store list --platform=armhf

FUNCTION                    DESCRIPTION
NodeInfo                    Get info about the machine that you'r...
Figlet                      Generate ASCII logos with the figlet CLI
SSL/TLS cert info           Returns SSL/TLS certificate informati...
nslookup                    Uses nslookup to return any IP addres...
Docker Image Manifest Query Query an image on the Docker Hub for ...
Left-Pad                    left-pad on OpenFaaS
```
```
./faas-cli-darwin store inspect nodeinfo --platform=armhf

FUNCTION DESCRIPTION                              IMAGE                           PROCESS      REPO
NodeInfo Get info about the machine that you'r... functions/nodeinfo:latest-armhf node main.js https://github.com/openfaas/faas/tree/master/sample-functions/NodeInfo
```


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

alexellis commented on Dockerfile.redist:
```diff
@@ -12,30 +12,36 @@ RUN test -z "$(gofmt -l $(find . -type f -name '*.go' -not -path "./vendor/*"))"
         -X github.com/openfaas/faas-cli/version.GitCommit=${GIT_COMMIT} \
-        -X github.com/openfaas/faas-cli/version.Version=${VERSION}" \
+        -X github.com/openfaas/faas-cli/version.Version=${VERSION} \
+        -X github.com/openfaas/faas-cli/platform.Platform=${PLATFORM}" \
```
I'm guessing that you need to hard-code the value for `PLATFORM` here or put it as a prefix to the `go build` command - otherwise won't the armhf binary have the wrong platform value as defined above?

alexellis replied:
What was your answer to this @viveksyngh ?

viveksyngh replied:
Now we are hardcoding the platform value for each type of platform like `x86_64 `,  `armhf` and `arm64`

alexellis commented on commands/store.go:
```diff
@@ -39,8 +43,9 @@ var storeCmd = &cobra.Command{
-	var results []schema.StoreItem
+func storeList(store, platform string) ([]schema.StoreFunction, error) {
+	var results []schema.StoreFunction
+	var storeData schema.StoreV2
```
@stefanprodan how does K8s like to do versioning in packages like this?

We had `Store` (v1) now we have Store v2

alexellis replied:
@stefanprodan ?

stefanprodan replied:
The version should be in the package name not in the struct see https://github.com/openfaas-incubator/openfaas-operator/tree/master/pkg/apis/openfaas

alexellis replied:
@viveksyngh can you address this? cc @rgee0

rgee0 commented:
Looking good @viveksyngh.

Can we get a view of the testing output for a platform that doesn't exist, please?  
I think it'd be good in this scenario if we could help the user by parsing the JSON - we have to fetch it anyway - and building a distinct slice of available platforms to feedback to the user.  wdyt?

Also, can we confirm that being case-sensitive is a conscious decision?  I know we've wandered into this minefield previously with the `dockerfile` template which took some unpicking.

viveksyngh commented:
Below is the output for the platform which does not exits. 

```
./faas-cli store list -p other
The store is empty.
```
I think providing the platform values as feedback in case of platform does not exists is a nice idea. 

 Platform input value are case sensitive. I think we should make it case insensitive for user input but consistency has to be maintained in the store json file.

rgee0 commented:
Lets change that message, something like:

`No items found in the store for platform 'other'`

If we can get the feedback in there then:

`No items found in the store for platform 'other', try one of the following: <output from parsing functions.json>`

Completely agree on the case sensitivity, followed a similar thought process when posing the earlier question.

viveksyngh commented:
Update this to provide platforms as feedback in case user adds a wrong platform and also added unit tests for new functions. 

```
./faas-cli-darwin store list

FUNCTION                    DESCRIPTION
NodeInfo                    Get info about the machine that you'r...
Figlet                      Generate ASCII logos with the figlet CLI
SSL/TLS cert info           Returns SSL/TLS certificate informati...
Colorization                Turn black and white photos to color ...
Inception                   This is a forked version of the work ...
Have I Been Pwned           The Have I Been Pwned function lets y...
Face Detection with Pigo    Detect faces in images using the Pigo...
SentimentAnalysis           Python function provides a rating on ...
Business Strategy Generator Generates a Business Strategy (using ...
Tesseract OCR               This function brings OCR - Optical Ch...
Dockerhub Stats             Golang function gives the count of re...
QR Code Generator - Go      QR Code generator using Go
Nmap Security Scanner       Tool for network discovery and securi...
ASCII Cows                  Generate cartoons of ASCII cows
YouTube Video Downloader    Download YouTube videos as a function
OpenFaaS Text-to-Speech     Generate an MP3 of text using Google'...
nslookup                    Uses nslookup to return any IP addres...
Docker Image Manifest Query Query an image on the Docker Hub for ...
face-detect with OpenCV     Detect faces in images. Send a URL as...
Left-Pad                    left-pad on OpenFaaS
mememachine                 Turn any image into a meme.

➜  faas-cli git:(new_store) ✗ ./faas-cli-darwin store list -p armhf

FUNCTION                    DESCRIPTION
NodeInfo                    Get info about the machine that you'r...
Figlet                      Generate ASCII logos with the figlet CLI
SSL/TLS cert info           Returns SSL/TLS certificate informati...
nslookup                    Uses nslookup to return any IP addres...
Docker Image Manifest Query Query an image on the Docker Hub for ...
Left-Pad                    left-pad on OpenFaaS

➜  faas-cli git:(new_store) ✗ ./faas-cli-darwin store list -p arm
No functions found in the store for platform 'arm', try one of the following: arm64, armhf, x86_64
➜  faas-cli git:(new_store) ✗ ./faas-cli-darwin store list -p arm64

FUNCTION  DESCRIPTION
NodeInfo  Get info about the machine that you'r...
sha512sum Generate shasums in 512 format

➜  faas-cli git:(new_store) ✗ ./faas-cli-darwin store inspect nodeinfo -p arm64

FUNCTION DESCRIPTION                              IMAGE                    PROCESS      REPO
NodeInfo Get info about the machine that you'r... functions/nodeinfo:arm64 node main.js https://github.com/openfaas/faas/tree/master/sample-functions/NodeInfo

➜  faas-cli git:(new_store) ✗ ./faas-cli-darwin store list -p arm
No functions found in the store for platform 'arm', try one of the following: armhf, x86_64, arm64
➜  faas-cli git:(new_store) ✗ ./faas-cli-darwin store inspect nodeinfo

FUNCTION DESCRIPTION                              IMAGE                     PROCESS      REPO
NodeInfo Get info about the machine that you'r... functions/nodeinfo:latest node main.js https://github.com/openfaas/faas/tree/master/sample-functions/NodeInfo

➜  faas-cli git:(new_store) ✗ ./faas-cli-darwin store deploy nodeinfo -p arm
Function 'nodeinfo' not found for platform 'arm'
➜  faas-cli git:(new_store) ✗ ./faas-cli-darwin store deploy nodeinfo
WARNING! Communication is not secure, please consider using HTTPS. Letsencrypt.org offers free SSL/TLS certificates.

Deployed. 202 Accepted.
URL: http://127.0.0.1:8080/function/nodeinfo
```

martindekov commented on commands/store_inspect.go:
```diff
@@ -32,31 +32,35 @@ func runStoreInspect(cmd *cobra.Command, args []string) error {
-	item := storeFindFunction(args[0], storeItems)
+	platformFunctions := filterStoreList(storeItems, targetPlatform)
+
+	functionName := args[0]
```
In case the naming is `ASCII Cows` this would fail

martindekov commented on commands/store.go:
```diff
@@ -23,12 +24,15 @@ var (
+
 func init() {
 	storeCmd.PersistentFlags().StringVarP(&storeAddress, "url", "u", defaultStore, "Alternative Store URL starting with http(s)://")
+	storeCmd.PersistentFlags().StringVarP(&platformValue, "platform", "p", "", "Target platform for store")
```
Should we put the default platform `defaultStore` here, instead of `""`, maybe work with `platformValue` directly? This way we can remove this package `platform/platform.go`

burtonr replied:
Adding a default value would make it more obvious when using the help commands as well

It could be added as a variable as well. With the current implementation, it would look like this:

`..., "p", platform.Platform, "Target platform...`

That would use the build parameter input and show as the default value in the help text:

```
Flags:
  -h, --help              help for store
  -p, --platform string   Target platform for store (default "x86_64")

```

alexellis replied:
@viveksyngh did you do this?

viveksyngh replied:
`platform` package has been removed and we are using `Platform` variable in `command` package to inject to value at build time and same value is being used as default value for the flag.

martindekov commented on commands/store_inspect.go:
```diff
@@ -32,31 +32,35 @@ func runStoreInspect(cmd *cobra.Command, args []string) error {
-	item := storeFindFunction(args[0], storeItems)
+	platformFunctions := filterStoreList(storeItems, targetPlatform)
+
+	functionName := args[0]
```
Can we do something for the times I directly write `ASCII Cows` without wrapping it in `''`? My motivation is primary UX here and I don't see `faas-cli store inspect` being used in different context/having different usage other than inspecting specific function. We can get all the arguments after `faas-cli store inspect` and cat them and then compare values.

alexellis replied:
Martin that is a good suggestion. Can you raise that as a separate issue?

@viveksyngh can you answer whether you addressed this or not? (I'm not saying that you have to, I just want to know what was done)

viveksyngh replied:
I also think this is very good suggestion somehow I have lost this comment. It is not addressed in this PR.

martindekov replied:
I will review it later today and will open Issue about this

alexellis replied:
@martindekov following up on this from June, can you take a look today?

martindekov replied:
Sorry for the delay with the answer Alex, I had [Issue](https://github.com/openfaas/faas-cli/issues/657) opened, but me and @LucasRoesler resolved it as unnecessary 👍

martindekov commented on platform/platform.go:
```diff
@@ -0,0 +1,18 @@
+)
+
+//GetPlatform get platform
+func GetPlatform() string {
```
If we set default value for the `--platform` flag, we can remove this functionality entirely

viveksyngh replied:
We have removed this package and we are using an default value for the flag.

martindekov commented on platform/platform.go:
```diff
@@ -0,0 +1,18 @@
+
+const (
+	//defaultPlatform platform x86_64
+	defaultPlatform = "x86_64"
```
You can use [mainPlatform](https://github.com/openfaas/faas-cli/blob/2487d1d3443c9b7201c98a2ecc57b617ac58db56/commands/template_store_list.go#L24) from the template store it is directly in the commands package so I believe you should be able to access it directly

viveksyngh replied:
We using this package to basically inject the value at the build time and use the same for that binary. If we use `mainPlatform` variable it will change the behavior of templates commands as well.

viveksyngh replied:
If that is something we want .. than We can use the same variable.

burtonr replied:
It would be good to consolidate the constants especially since they are used in the same context with the function store.Perhaps update `mainPlatform` to `defaultPlatform` and use that one constant everywhere.

This will make maintenance much easier in the future.

alexellis replied:
What did we decide on?

viveksyngh replied:
We have removed `platform` package and using an variable in `commands` package name `Platform` which is injected at the build time. if that value is empty then we are using `mainPlatform` as default value for the platform.

martindekov commented on platform/platform.go:
```diff
@@ -0,0 +1,18 @@
+package platform
```
I believe we can go without adding this package

alexellis replied:
@viveksyngh please can you answer this?

viveksyngh replied:
`platform` package has been remove and we are using `mainPlatform` variable for default value.

burtonr commented on commands/store.go:
```diff
@@ -31,8 +43,63 @@ var storeCmd = &cobra.Command{
+	var filteredList []schema.StoreFunction
+
+	for _, function := range functions {
+		_, ok := function.Images[platform]
```
This needs to be updated to not be case sensitive. Perhaps force the `platform` input to lower case?
Or, change this function to loop and compare with `strings.EqualFold()`...

alexellis replied:
@viveksyngh ?

viveksyngh replied:
Yes this been taken care while comparing the value and it is updated in the test results.

burtonr commented:
@viveksyngh Added some comments that would be some potential improvements to the great work that has already been done. I don't mean to take away from the work, just suggestions. 

I was able to test it out locally, and everything worked exactly as expected. The only surprise was the case-sensitive parameter, which is not really a deal-breaker. 

Nice work! :100:

viveksyngh commented:
I have updated the PR to address review comments. Please have a look.

alexellis commented on schema/store_item.go:
```diff
@@ -15,3 +15,31 @@ type StoreItem struct {
 	ReadOnlyRootFilesystem bool              `json:"readOnlyRootFilesystem"`
 }
+
+//StoreFunction function metadata in store
```
@viveksyngh please answer the comment

viveksyngh replied:
I have updated this

alexellis commented:
Vivek thank you for leading this task.

We have so many unanswered comments here that as a maintainer, I have no idea what has been done or whether the comments have even been acknowledged.

That is not going to work at this scale, please can we all try to acknowledge all comments by replying? Not replying to a comment, but changing the code makes it impossible for me to know what has happened.

@viveksyngh please could you answer as many of the questions as is practically possible.

@burtonr, @martindekov and @rgee0 please can you re-review and if necessary raise issues for the comment about typing in "ASCII Cows" without quotes.

alexellis commented:
I would also like to know where we got to with regards to the `platform` package.

alexellis commented:
/set title: Use functions.json file for store

viveksyngh commented on commands/store_list.go:
```diff
@@ -31,22 +31,27 @@ var storeListCmd = &cobra.Command{
 
-	if len(items) == 0 {
-		fmt.Printf("The store is empty.")
+	availablePlatforms := getStorePlatforms(storeList)
```
This has been also updated.

viveksyngh commented:
Yes I think I made mistakes of not answering each questions individually but I have addressed most of the comments in my last commit and I commented that. Now I have replied to reach individual comment. 

Platform package has been removed and making use of variable defined in the commands package.

  "ASCII Cows" issue has not be addressed in this PR.

alexellis commented on schema/store_item.go:
```diff
@@ -15,3 +15,31 @@ type StoreItem struct {
 }
+
+//StoreFunction represents a multi-arch function in the store
+type StoreFunction struct {
```
Broader conversation here, @LucasRoesler @stefanprodan would this be useful in faas-provider?

stefanprodan replied:
I think StoreFunction could embed the Function type like:

```go
import "github.com/openfaas/faas-provider/v1"

type StoreFunction struct {
	v1.Function
	Icon                   string            `json:"icon"`
	RepoURL          string            `json:"repo_url"`
}
```

LucasRoesler replied:
The primary benefit I can see of including it in the faas-provider would be for exporting/reconstructing the definition.  Is this important that all providers can return the original store function definition?

alexellis replied:
Thanks for the input, let's come back to this after the PR is merged?

martindekov commented:
Raised an [Issue](https://github.com/openfaas/faas-cli/issues/657) for the Cows, not sure if this is relevant about the template store.

alexellis commented:
Hi @viveksyngh could you spend a bit of timing looking over this PR to see if it's something we can merge, or whether it needs more work?

I think it would be good to have available sooner than later.

A rebase is also needed on the store_deploy.go file

Alex

alexellis commented:
Due date for the milestone is Friday 1st Nov. Does that work for you @viveksyngh ?

viveksyngh commented:
I have addressed all the comments along with one moving version to package rather than in struct name from @stefanprodan. It is ready for review and if changes required I can complete before friday.

alexellis commented:
Thank you to everyone who was involved.
